compiler/rustc_ast/src/ast.rs

```diff
@@ -692,10 +692,10 @@ pub enum PatKind {
 
     /// A struct or struct variant pattern (e.g., `Variant {x, y, ..}`).
     /// The `bool` is `true` in the presence of a `..`.
-    Struct(Path, Vec<FieldPat>, /* recovered */ bool),
+    Struct((Option<QSelf>, Path), Vec<FieldPat>, /* recovered */ bool),
```

Comment:
```suggestion
    Struct(Option<QSelf>, Path, Vec<FieldPat>, /* recovered */ bool),
```
Could you flatten all these weird tuples in AST?

The issue with size of `ExprKind` is addressed in https://github.com/rust-lang/rust/pull/83188 by boxing all its fields instead of this tuple.